dali/pipeline/operators/decoder/nvjpeg_decoder_gpu.h

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_DECODER_NVJPEG_DECODER_GPU_H_
+#define DALI_PIPELINE_OPERATORS_DECODER_NVJPEG_DECODER_GPU_H_
+
+#include <functional>
+#include <utility>
+#include <vector>
+
+#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operators/decoder/nvjpeg_helper.h"
+#include "dali/util/ocv.h"
+
+namespace dali {
+
+using ImageInfo = EncodedImageInfo<unsigned int>;
+
+class nvJPEGDecoderGPUStage : public Operator<MixedBackend> {
+ public:
+  explicit nvJPEGDecoderGPUStage(const OpSpec& spec) :
+    Operator<MixedBackend>(spec),
+    output_image_type_(spec.GetArgument<DALIImageType>("output_type")) {
+    NVJPEG_CALL(nvjpegCreateSimple(&handle_));
+
+    NVJPEG_CALL(nvjpegDecoderCreate(handle_, NVJPEG_BACKEND_HYBRID, &decoder_host_));
+    NVJPEG_CALL(nvjpegDecoderCreate(handle_, NVJPEG_BACKEND_GPU_HYBRID, &decoder_hybrid_));
+
+    NVJPEG_CALL(nvjpegBufferDeviceCreate(handle_, nullptr, &device_buffer_));
+
+    // Do we really need both in both stages ops?
+    size_t device_memory_padding = spec.GetArgument<Index>("device_memory_padding");
+    size_t host_memory_padding = spec.GetArgument<Index>("host_memory_padding");
+    NVJPEG_CALL(nvjpegSetDeviceMemoryPadding(device_memory_padding, handle_));
+    NVJPEG_CALL(nvjpegSetPinnedMemoryPadding(host_memory_padding, handle_));
+  }
+
+  ~nvJPEGDecoderGPUStage() noexcept(false) {
+    NVJPEG_CALL(nvjpegBufferDeviceDestroy(device_buffer_));
+    NVJPEG_CALL(nvjpegDecoderDestroy(decoder_host_));
+    NVJPEG_CALL(nvjpegDecoderDestroy(decoder_hybrid_));
+    NVJPEG_CALL(nvjpegDestroy(handle_));
+  }
+
+  using dali::OperatorBase::Run;
+  void Run(MixedWorkspace *ws) override {
+    std::vector<Dims> output_shape(batch_size_);
+    // Creating output shape and setting the order of images so the largest are processed first
+    // (for load balancing)
+    std::vector<std::pair<size_t, size_t>> image_order(batch_size_);
+    for (int i = 0; i < batch_size_; i++) {
+      const auto& info_tensor = ws->Input<CPUBackend>(0, i);
+      const auto& state_tensor = ws->Input<CPUBackend>(1, i);
+      const ImageInfo* info;
+      std::tie(info, std::ignore) = GetInfoState(info_tensor, state_tensor);
```

Comment:
This access to state_tensor and reinterpreting it here is uncessary as it is unused.

Reply:
Agree, done